File: apps/mobile/lib/services/financial_core/money_truth_receipt.dart
// ────────────────────────────────────────────────────────────
//  MoneyTruthReceipt v1 — contrat de vérité chiffrée (Tranche firstJob PR-B)
//
//  Miroir Dart du modèle backend
//  `services/backend/app/models/lucidity/money_truth_receipt.py`.
//  Le receipt ENVELOPPE une valeur L1 (net firstJob, offline-capable) avec sa
//  provenance : inputs normalisés, hash déterministe, millésime, sources
//  datées, bande d'incertitude, confiance.
//
//  Parité (SPEC §4.1 / §4.4) : pour les MÊMES inputs, ce miroir produit le
//  MÊME `inputsHash` (octet pour octet) et la MÊME `value` après arrondi que
//  le producteur backend. Verrou : tools/fixtures/money_truth_receipt_v1.json
//  (rejoué par pytest ET flutter test).
//
//  `computeInputsHash` réutilise l'algorithme prouvé du harnais de parité
//  `apps/mobile/tools/hash_parity_harness/main.dart` (quantize-BEFORE-
//  canonicalize, RFC 8785, SHA256), lui-même validé octet pour octet contre
//  `app.services.coach.inputs_hash.compute_inputs_hash` par
//  `services/backend/tests/test_dag_invalidation/test_hash_parity.py`.
//
//  Interdit (extra=forbid côté backend) : tout champ de classement
//  (recommended* / best* / optimal* …). Aucun ici — cohérent NEVER #5 / LSFin.
// ────────────────────────────────────────────────────────────

import 'dart:convert';

import 'package:crypto/crypto.dart';

/// Claim id du producteur v1 (SPEC D3 / §4.1).
const String kFirstJobNetSalaryClaimId = 'firstjob.net_salary.v1';

// ════════════════════════════════════════════════════════════
//  HASH DÉTERMINISTE DES INPUTS (miroir de compute_inputs_hash)
// ════════════════════════════════════════════════════════════

/// Quantifie récursivement les doubles à 2 décimales (ROUND_HALF_UP).
///
/// Miroir EXACT de `_quantize_floats()` (inputs_hash.py) et de `_quantize()`
/// (hash_parity_harness/main.dart). L'ordre de récursion DOIT correspondre à
/// Python sinon les hashes divergent.
dynamic _quantize(dynamic v) {
  if (v is Map) {
    final out = <String, dynamic>{};
    v.forEach((k, val) {
      out[k as String] = _quantize(val);
    });
    return out;
  }
  if (v is List) {
    return v.map(_quantize).toList();
  }
  if (v is bool) return v;
  if (v is int) return v;
  if (v is double) {
    if (v.isNaN) {
      throw const FormatException('NaN not allowed in projection inputs');
    }
    if (!v.isFinite) {
      throw const FormatException('inf/-inf not allowed in projection inputs');
    }
    // Decimal(0.01) ROUND_HALF_UP : ×100, round-half-away-from-zero, ÷100.
    final scaled = v * 100.0;
    final rounded = scaled.roundToDouble();
    return rounded / 100.0;
  }
  return v;
}

/// JSON canonique RFC 8785 (clés triées UTF-16, sans espaces, floats entiers
/// sans `.0`). Miroir EXACT de `canonicalize()` du harnais.
String _canonicalize(dynamic v) {
  if (v is Map) {
    final keys = v.keys.cast<String>().toList()..sort();
    return '{${keys.map((k) => '${jsonEncode(k)}:${_canonicalize(v[k])}').join(',')}}';
  }
  if (v is List) {
    return '[${v.map(_canonicalize).join(',')}]';
  }
  if (v is double) {
    if (v.truncateToDouble() == v && v.isFinite) {
      return v.toInt().toString();
    }
    return v.toString();
  }
  if (v is bool) return v ? 'true' : 'false';
  if (v == null) return 'null';
  return jsonEncode(v); // strings + ints
}

/// SHA256 hex (64 car.) du JSON canonique des inputs quantifiés.
///
/// Déterministe et identique octet pour octet à
/// `compute_inputs_hash(inputs)` côté Python.
String computeInputsHash(Map<String, dynamic> inputs) {
  final quantized = _quantize(inputs);
  final canon = _canonicalize(quantized);
  return sha256.convert(utf8.encode(canon)).toString();
}

// ════════════════════════════════════════════════════════════
//  VALUE OBJECTS
// ════════════════════════════════════════════════════════════

/// Une source datée derrière la valeur (SPEC §4.1 `sources[]`).
class MoneyTruthSource {
  final String id;
  final String label;
  final int vintage;

  const MoneyTruthSource({
    required this.id,
    required this.label,
    required this.vintage,
  });

  Map<String, dynamic> toJson() => {
        'id': id,
        'label': label,
        'vintage': vintage,
      };
}

/// Bande d'incertitude {low, high} (SPEC §4.1 `range`).
class MoneyTruthRange {
  final double low;
  final double high;

  const MoneyTruthRange({required this.low, required this.high});

  Map<String, dynamic> toJson() => {'low': low, 'high': high};
}

/// Résumé de confiance — MÊME forme fil que `EnhancedConfidence`
/// (confidence_scorer.dart:995 / app/schemas/enhanced_confidence.py) : 4 axes
/// [0,1] + score combiné. PAS un nouveau scorer (règle 4 / NEVER #3) : juste
/// la forme fil peuplée par le mapping déclaré PR-B.
class MoneyTruthConfidence {
  final double completeness;
  final double accuracy;
  final double freshness;
  final double understanding;
  final double score;

  const MoneyTruthConfidence({
    required this.completeness,
    required this.accuracy,
    required this.freshness,
    required this.understanding,
    required this.score,
  });

  Map<String, dynamic> toJson() => {
        'completeness': completeness,
        'accuracy': accuracy,
        'freshness': freshness,
        'understanding': understanding,
        'score': score,
      };
}

/// Contrat de vérité chiffrée v1 — enveloppe une valeur L1 + sa provenance.
///
/// Immuable (tous champs `final`). Sérialisation camelCase alignée octet sur
/// le backend Pydantic v2.
class MoneyTruthReceipt {
  final String claimId;
  final String receiptId;
  final Map<String, dynamic> inputs;
  final String inputsHash;
  final String jurisdiction;
  final int taxYear;
  final String base; // "brut" | "net"
  final String civilStatus;
  final List<String> assumptions;
  final String engine;
  final String engineVersion;
  final String rounding;
  final List<MoneyTruthSource> sources;
  final double value;
  final MoneyTruthRange? range;
  final MoneyTruthConfidence? confidence;
  final String computedAt;

  const MoneyTruthReceipt({
    required this.claimId,
    required this.receiptId,
    required this.inputs,
    required this.inputsHash,
    required this.jurisdiction,
    required this.taxYear,
    required this.base,
    required this.civilStatus,
    required this.assumptions,
    required this.engine,
    required this.engineVersion,
    required this.rounding,
    required this.sources,
    required this.value,
    required this.range,
    required this.confidence,
    required this.computedAt,
  });

  Map<String, dynamic> toJson() => {
        'claimId': claimId,
        'receiptId': receiptId,
        'inputs': inputs,
        'inputsHash': inputsHash,
        'jurisdiction': jurisdiction,
        'taxYear': taxYear,
        'base': base,
        'civilStatus': civilStatus,
        'assumptions': assumptions,
        'engine': engine,
        'engineVersion': engineVersion,
        'rounding': rounding,
        'sources': sources.map((s) => s.toJson()).toList(),
        'value': value,
        'range': range?.toJson(),
        'confidence': confidence?.toJson(),
        'computedAt': computedAt,
      };
}

File: services/backend/app/models/lucidity/money_truth_receipt.py
"""MoneyTruthReceipt v1 — contrat de vérité chiffrée (Tranche firstJob PR-B).

Ce modèle ENVELOPPE une valeur L1 (single-number, offline-capable) avec sa
provenance : inputs normalisés, hash déterministe des inputs, millésime,
sources datées, bande d'incertitude et confiance. Il ne remplace PAS le
discriminateur `_payload.py` (L1ChiffrePayload …) ; il l'enrichit — le receipt
accompagne la valeur dans l'enveloppe `CoachToolOk.data` (SPEC §4.2).

Réutilisation obligatoire (SPEC §4.2, pas de réinvention) :
- `inputs_hash` ← `app.services.coach.inputs_hash.compute_inputs_hash`
  (SHA256 hex de JSON canonique RFC 8785, floats quantifiés Decimal(0.01)).
  Le miroir Dart (`money_truth_receipt.dart`) porte le MÊME algorithme, validé
  octet pour octet par `tools/fixtures/money_truth_receipt_v1.json`.
- `confidence` ← forme fil `EnhancedConfidence`
  (`app.schemas.enhanced_confidence`, mêmes 4 axes que
  `confidence_scorer.dart:995`) — pas de nouveau scorer (NEVER #3 / règle 4).
- Persistance (PR-E, hors de ce fichier) ← `ProjectionAuditRecord`
  (`inputs_hash → scenario_inputs_hash`, `hash(value) → output_hash`,
  `engine_version → constants_version_hash + app_version`).

Doctrine de type (miroir de `_payload.py`, SPEC §4.1) :
- `extra="forbid"` : AUCUN champ inconnu ne peut être ajouté. Les champs de
  classement (`recommended*` / `best*` / `top_pick` / `preferred` / `winning*`
  et paraphrases) n'ont donc PAS de nom de champ à peupler — impossibilité de
  schéma (cohérent NEVER #5 / LSFin art. 8).
- `frozen=True` : aucune mutation post-construction (pas d'altération de la
  valeur après validation).
- Sérialisation camelCase (`alias_generator=to_camel`, `populate_by_name=True`)
  pour l'alignement octet avec le miroir Dart et le transport Pydantic v2.

Producteur v1 : le claim `firstjob.net_salary.v1`, émis par
`app.services.first_job.onboarding_service` (le net firstJob backend produit le
receipt). Le receipt v1 reste INTERNE (pas exposé via OpenAPI ; le portage API
= PR-E).

Références
----------
- .planning/phases/mint-2-0-first-experience-rente-capital/TRANCHE-FIRSTJOB-SPEC.md
  §4 (schéma 16 champs, §4.1 exigences range/confidence/taxYear/fixtures),
  §6 (découpage PR-B).
- app/models/lucidity/_payload.py (doctrine extra=forbid + frozen).
- app/services/coach/inputs_hash.py (compute_inputs_hash réutilisé).
"""
from __future__ import annotations

from types import MappingProxyType
from typing import Any, Literal, Mapping, Optional, Tuple

from pydantic import (
    BaseModel,
    ConfigDict,
    Field,
    field_serializer,
    field_validator,
    model_validator,
)
from pydantic.alias_generators import to_camel

from app.schemas.enhanced_confidence import EnhancedConfidence

# ---------------------------------------------------------------------------
# Claim id du producteur v1 (SPEC D3 / §4.1). Le seul claim câblé en PR-B.
# ---------------------------------------------------------------------------

FIRST_JOB_NET_SALARY_CLAIM_ID = "firstjob.net_salary.v1"


class _MoneyTruthBase(BaseModel):
    """Config partagée : forbid extras + immuable + camelCase (SPEC §4.1)."""

    model_config = ConfigDict(
        extra="forbid",
        frozen=True,
        populate_by_name=True,
        alias_generator=to_camel,
    )


class MoneyTruthSource(_MoneyTruthBase):
    """Une source datée derrière la valeur (SPEC §4.1 `sources[]`).

    `vintage` = millésime de la source (ex. 2026 pour les taux de cotisation,
    2022 pour une statistique OFS ESS). Ne JAMAIS confondre avec `tax_year`
    (millésime des règles de cotisation appliquées) — SPEC §4.1 précision.
    """

    id: str = Field(..., min_length=1)
    label: str = Field(..., min_length=1)
    vintage: int = Field(..., ge=1900, le=2100)


class MoneyTruthRange(_MoneyTruthBase):
    """Bande d'incertitude {low, high} sur la valeur (SPEC §4.1 `range`).

    Chaque borne est un RECALCUL du service avec une hypothèse bornante
    basse / haute (PAS un pourcentage arbitraire) — SPEC §4.1 précision.
    """

    low: float
    high: float

    @model_validator(mode="after")
    def _low_le_high(self) -> "MoneyTruthRange":
        if self.low > self.high:
            raise ValueError(
                f"MoneyTruthRange invalide : low ({self.low}) > high ({self.high})."
            )
        return self


class MoneyTruthReceipt(_MoneyTruthBase):
    """Contrat de vérité chiffrée v1 — enveloppe une valeur L1 + sa provenance.

    Champs (SPEC §4.1). `range` et `confidence` sont nullables EN GÉNÉRAL mais
    REQUIS pour `claim_id == firstjob.net_salary.v1` (validateur ci-dessous ;
    A3 les affiche).
    """

    claim_id: str = Field(..., min_length=1)
    receipt_id: str = Field(..., min_length=1)
    # Immutabilité PROFONDE (frozen=True n'est que superficiel) : les
    # collections sont gelées à la construction — `inputs` en MappingProxyType,
    # `assumptions`/`sources` en tuples — pour que la valeur et sa provenance ne
    # puissent pas être altérées après validation (tamper-evidence).
    inputs: Mapping[str, Any]
    inputs_hash: str = Field(..., min_length=64, max_length=64)
    jurisdiction: str = Field(..., min_length=1)
    tax_year: int = Field(..., ge=1900, le=2100)
    base: Literal["brut", "net"]
    civil_status: str = Field(..., min_length=1)
    assumptions: Tuple[str, ...]
    engine: str = Field(..., min_length=1)
    engine_version: str = Field(..., min_length=1)
    rounding: str = Field(..., min_length=1)
    sources: Tuple[MoneyTruthSource, ...] = Field(..., min_length=1)
    value: float
    range: Optional[MoneyTruthRange] = None
    confidence: Optional[EnhancedConfidence] = None
    computed_at: str = Field(..., min_length=1)

    @field_validator("inputs", mode="after")
    @classmethod
    def _freeze_inputs(cls, v: Mapping[str, Any]) -> Mapping[str, Any]:
        """Gèle `inputs` en vue lecture-seule (mutation post-construction -> erreur)."""
        return MappingProxyType(dict(v))

    @field_serializer("inputs")
    def _serialize_inputs(self, v: Mapping[str, Any], _info: Any) -> dict:
        """Sérialise la MappingProxyType en dict JSON standard."""
        return dict(v)

    @model_validator(mode="after")
    def _net_salary_requires_band_and_confidence(self) -> "MoneyTruthReceipt":
        """`range` + `confidence` REQUIS pour firstjob.net_salary.v1 (SPEC §4.1).

        L'appareil de lucidité (bande + confiance) ne peut pas manquer sur le
        premier chiffre — le rendre nullable pour ce claim casserait A3.
        """
        if self.claim_id == FIRST_JOB_NET_SALARY_CLAIM_ID:
            if self.range is None:
                raise ValueError(
                    f"claim {FIRST_JOB_NET_SALARY_CLAIM_ID} exige un `range` "
                    "(bande d'incertitude) non nul — SPEC §4.1 / A3."
                )
            if self.confidence is None:
                raise ValueError(
                    f"claim {FIRST_JOB_NET_SALARY_CLAIM_ID} exige une "
                    "`confidence` non nulle — SPEC §4.1 / A3."
                )
        return self


__all__ = [
    "FIRST_JOB_NET_SALARY_CLAIM_ID",
    "MoneyTruthReceipt",
    "MoneyTruthRange",
    "MoneyTruthSource",
]
